Encoder/DefaultEncoder.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+namespace Lexik\Bundle\JWTAuthenticationBundle\Encoder;
+
+use InvalidArgumentException;
+use Lexik\Bundle\JWTAuthenticationBundle\Services\JWSProviderInterface;
+use Lexik\Bundle\JWTAuthenticationBundle\Exception\JWTDecodeFailure\ExpiredJWTDecodeFailureException;
+use Lexik\Bundle\JWTAuthenticationBundle\Exception\JWTEncodeFailure\JWTEncodeFailureException;
+use Lexik\Bundle\JWTAuthenticationBundle\Exception\JWTDecodeFailure\JWTDecodeFailureException;
+use Lexik\Bundle\JWTAuthenticationBundle\Exception\JWTDecodeFailure\UnverifiedJWTDecodeFailureException;
+use Lexik\Bundle\JWTAuthenticationBundle\Exception\JWTEncodeFailure\UnsignedJWTEncodeFailureException;
+
+/**
+ * Default Json Web Token encoder/decoder.
+ *
+ * @since 2.0
+ *
+ * @author Robin Chalas <robin.chalas@gmail.com>
+ */
+class DefaultEncoder extends JWTEncoder
+{
+    /**
+     * @var JWSProviderInterface
+     */
+    protected $jwsProvider;
+
+    /**
+     * @param JWSProviderInterface $jwsProvider
+     */
+    public function __construct(JWSProviderInterface $jwsProvider)
+    {
+        $this->jwsProvider = $jwsProvider;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws UnsignedJWTEncodeFailureException
+     */
+    public function encode(array $data)
+    {
+        try {
+            $this->jwsProvider->createSignedToken($data);
+        } catch (InvalidArgumentException $e) {
+            throw new JWTEncodeFailureException('An error occurred while trying to encode the JWT token.', $e);
+        }
+
+        if (JWSProviderInterface::SIGNED !== $this->jwsProvider->getStatus()) {
+            throw new UnsignedJWTEncodeFailureException();
+        }
+
+        return $this->jwsProvider->getToken();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws JWTDecodeFailureException           If the signature cannot be loaded
+     * @throws UnverifiedJWTDecodeFailureException If the signature cannot be verified
+     * @throws ExpiredJWTDecodeFailureException    If the token is expired
```

Comment:
All of the details about the exceptions should be moved to the interface. This is one of the "BC" breaks we need to add to an UPGRADE file. Specifically, both `encode()` and `decode()` previously did *not* throw exceptions on failures, but now they do. We "changed" how the interface behaves :) (it's a good change!). But if you're upgrading, you'll need to know about this. For example, in my KnpU tutorial, we are checking to see if the result of decode is `false` - now we'll need to catch the exceptions.